File: Runtime/Plugins/Android/OnFetchCompleteListenerProxy.java
package com.yandex.varioqub.plugin.unity;

import androidx.annotation.NonNull;

import com.yandex.varioqub.config.FetchError;
import com.yandex.varioqub.config.OnFetchCompleteListener;

public interface OnFetchCompleteListenerProxy {

    void onError(@NonNull String reason);

    void onSuccess();

    class Delegate implements OnFetchCompleteListener {

        private final OnFetchCompleteListenerProxy proxy;

        public Delegate(OnFetchCompleteListenerProxy proxy) {
            this.proxy = proxy;
        }

        @Override
        public void onError(@NonNull String s, @NonNull FetchError fetchError) {
            proxy.onError(fetchErrorToString(fetchError));
        }

        @Override
        public void onSuccess() {
            proxy.onSuccess();
        }

        private String fetchErrorToString(FetchError error) {
            switch (error) {

                case EMPTY_RESULT:
                    return "EMPTY_RESULT";
                case IDENTIFIERS_NULL:
                    return "IDENTIFIERS_NULL";
                case RESPONSE_PARSE_ERROR:
                    return "RESPONSE_PARSE_ERROR";
                case REQUEST_THROTTLED:
                    return "REQUEST_THROTTLED";
                case NETWORK_ERROR:
                    return "NETWORK_ERROR";
                case INTERNAL_ERROR:
                    return "INTERNAL_ERROR";
                default:
                    return "";
            }
        }
    }
}

File: Runtime/Native/Android/Proxy/OnFetchCompleteListenerProxy.cs
#if UNITY_ANDROID
using JetBrains.Annotations;
using UnityEngine;

namespace Com.Yandex.Varioqub.Native.Android.Proxy {
    internal class OnFetchCompleteListenerProxy : AndroidJavaProxy {
        [NotNull]
        private readonly OnFetchCompleteListener.OnSuccessDelegate _onSuccessDelegate;
        [CanBeNull]
        private readonly OnFetchCompleteListener.OnErrorDelegate _onErrorDelegate;

        public OnFetchCompleteListenerProxy(
            [NotNull] OnFetchCompleteListener.OnSuccessDelegate onSuccessDelegate,
            [CanBeNull] OnFetchCompleteListener.OnErrorDelegate onErrorDelegate
        ) : base("com.yandex.varioqub.plugin.unity.OnFetchCompleteListenerProxy") {
            _onSuccessDelegate = onSuccessDelegate;
            _onErrorDelegate = onErrorDelegate;
        }

        void onSuccess() {
            _onSuccessDelegate();
        }

        void onError([NotNull] string error) {
            _onErrorDelegate?.Invoke(GetError(error));
        }

        private static OnFetchCompleteListener.Error GetError([NotNull] string str) {
            return str switch {
                "EMPTY_RESULT" => OnFetchCompleteListener.Error.EmptyResult,
                "IDENTIFIERS_NULL" => OnFetchCompleteListener.Error.IdentifiersNull,
                "RESPONSE_PARSE_ERROR" => OnFetchCompleteListener.Error.ResponseParseError,
                "REQUEST_THROTTLED" => OnFetchCompleteListener.Error.RequestThrottled,
                "NETWORK_ERROR" => OnFetchCompleteListener.Error.NetworkError,
                "INTERNAL_ERROR" => OnFetchCompleteListener.Error.InternalError,
                _ => OnFetchCompleteListener.Error.Unknown,
            };
        }
    }
}
#endif
